lens/vector/repo.go

```diff
@@ -0,0 +1,179 @@
+package vector
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/urfave/cli/v2"
+
+	"github.com/ipfs/go-blockservice"
+	cid "github.com/ipfs/go-cid"
+	offline "github.com/ipfs/go-ipfs-exchange-offline"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	car "github.com/ipld/go-car"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/vm"
+	"github.com/filecoin-project/lotus/extern/sector-storage/ffiwrapper"
+	"github.com/filecoin-project/lotus/journal"
+	"github.com/filecoin-project/lotus/lib/bufbstore"
+	"github.com/filecoin-project/lotus/lib/ulimit"
+	"github.com/filecoin-project/lotus/node/impl"
+	"github.com/filecoin-project/lotus/node/impl/full"
+	"github.com/filecoin-project/lotus/node/repo"
+	"github.com/filecoin-project/sentinel-visor/lens"
+	"github.com/filecoin-project/sentinel-visor/lens/util"
+	"github.com/filecoin-project/specs-actors/actors/runtime/proof"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+)
+
+func NewAPIOpener(c *cli.Context) (*APIOpener, lens.APICloser, error) {
+	capi := CaptureAPI{}
+
+	if _, _, err := ulimit.ManageFdLimit(); err != nil {
+		return nil, nil, fmt.Errorf("setting file descriptor limit: %s", err)
+	}
+
+	r, err := repo.NewFS(c.String("repo"))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	exists, err := r.Exists()
+	if err != nil {
+		return nil, nil, err
+	}
+	if !exists {
+		return nil, nil, fmt.Errorf("lotus repo doesn't exist")
+	}
+
+	var lr repo.LockedRepo
+	if c.Bool("repo-read-only") {
+		lr, err = r.LockRO(repo.FullNode)
+	} else {
+		lr, err = r.Lock(repo.FullNode)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sf := func() {
+		lr.Close()
+	}
+
+	bs, err := lr.Blockstore(repo.BlockstoreChain)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// wrap the repos blockstore with a tracing implementation capturing all cid's read.
+	capi.tbs = NewTracingBlockstore(bs)
```

Comment:
Wrapping the API's blockstore with a `TracingBlockstore` tracks CIDs read from it. The recorded CIDs are used to build the vectors CAR data.